Support for TPUv5 and v6e

wiktorn opened the pull request:
Add support for TPU v5e and V6e.

* add support for startup scripts on TPU nodes
* breaking change: rename `tf_version` to `runtime_version`
* add support for spot TPU nodes
* reduced number of TPU API calls by slurmsync.py
* include startup script in `slurm-gcp-devel.zip` - otherwise only version from image is used

TODO:
* align on the naming of Docker containers in the public repository

Linked PR: https://github.com/GoogleCloudPlatform/slurm-gcp/pull/265
Obsoletes: #3927

wiktorn committed:
Add support for TPU V5e, V5p and V6e

wiktorn committed:
Reduce number of TPU API calls

wiktorn committed:
Add support for startup scripts on TPU nodesets

mr0re1 commented on community/modules/internal/slurm-gcp/instance_template/files/startup_sh_unlinted:
```diff
@@ -14,6 +14,7 @@
 # limitations under the License.
 
 set -e
+set -o xtrace
```
This file is called `unlinted` to workaround shell-linter that forced us to make changes to this file content.
We wanted to avoid such changes as it would force replacement of templates. I see few options here:
* Embrace breaking change: rename it and fix all lint error + your changes;
* Lazily embrace breaking change: what you have in your PR (totally fine);
* Make it non-breaking: create a separate startup-script for TPUs, keep non-TPU as-is.

WDYT?

wiktorn replied:
Went with:
> Make it non-breaking: create a separate startup-script for TPUs, keep non-TPU as-is.

mr0re1 commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/main.tf:
```diff
@@ -171,21 +171,41 @@ locals {
+  ])
+}
+
+data "local_file" "startup_sh" {
```
Why does it needs to be in devel ?

wiktorn replied:
This is currently [referred locally](https://github.com/GoogleCloudPlatform/cluster-toolkit/blob/main/community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tpu.py#L216-L219) when creating a TPU node. I'm adding it here for a sake of completeness and improve development. Otherwise you need to rebuild controller image each time you change this script (or do the change manually).

TPUs do not have yet any templates to keep them there :-/

wiktorn replied:
With a dedicated script, there are no changes needed here.

mr0re1 commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/requirements.txt:
```diff
@@ -7,7 +7,7 @@ google-cloud-bigquery==3.11.3
 google-cloud-secret-manager~=2.22
 google-cloud-storage==2.10.0
-google-cloud-tpu==1.10.0
+google-cloud-tpu==1.21.0
```
Sadly, that will not be enough, we need to change requirements.txt in slurm-gcp repo and roll an image first.

wiktorn replied:
Just a note, slurm-gcp repo uses lower versions for following packages:
```
requests==2.31.0
PyYAML==6.0.1
```

wiktorn committed:
Align versions to slurm-gcp requirements.txt

Bumped versions to the versions defined in slurm-gcp repository.

Did not downgrade:
requests==2.31.0
PyYAML==6.0.1

wiktorn committed:
Review fixes

* use separate startup script for TPU nodes
* fix container references in README
* document list_nodes and fix caching of multi-page responses

wiktorn committed:
Add OS_ENV to the container, to ensure that container is properly detected

wiktorn commented:
Waiting for external changes

Neelabh94 commented:
@wiktorn Closing this since there have been no activity on it for quite some time. Please reopen if anything changes. 
Thanks.
